sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -1582,6 +1618,131 @@ impl TypedExpression {
+                            );
+                            if !variable_decl.is_mutable.is_mutable() {
+                                errors.push(CompileError::AssignmentToNonMutable { name });
+                                return err(warnings, errors);
+                            }
+                            break (name, variable_decl.body.return_type);
+                        }
+                        ExpressionKind::Subfield(SubfieldExpression {
+                            prefix,
+                            field_to_access,
+                            ..
+                        }) => {
+                            names_vec.push(ProjectionKind::StructField {
+                                name: field_to_access,
+                            });
+                            expr = prefix;
+                        }
+                        ExpressionKind::TupleIndex(TupleIndexExpression {
+                            prefix,
+                            index,
+                            index_span,
+                            ..
+                        }) => {
+                            names_vec.push(ProjectionKind::TupleField { index, index_span });
+                            expr = prefix;
+                        }
+                        _ => {
+                            errors.push(CompileError::InvalidExpressionOnLhs { span });
+                            return err(warnings, errors);
+                        }
+                    }
+                };
+                let names_vec = names_vec.into_iter().rev().collect::<Vec<_>>();
+                let (ty_of_field, _ty_of_parent) = check!(
+                    ctx.namespace.find_subfield_type(&base_name, &names_vec),
+                    return err(warnings, errors),
+                    warnings,
+                    errors
+                );
+                // type check the reassignment
+                let ctx = ctx.with_type_annotation(ty_of_field).with_help_text("");
+                let rhs_span = rhs.span();
+                let rhs = check!(
+                    TypedExpression::type_check(ctx, rhs),
+                    error_recovery_expr(rhs_span),
+                    warnings,
+                    errors
+                );
+
+                ok(
+                    TypedExpression {
+                        expression: TypedExpressionVariant::Reassignment(Box::new(
+                            TypedReassignment {
+                                lhs_base_name: base_name,
+                                lhs_type: final_return_type,
+                                lhs_indices: names_vec,
+                                rhs,
+                            },
+                        )),
+                        return_type: crate::type_system::insert_type(TypeInfo::Tuple(Vec::new())),
+                        // TODO: if the rhs is constant then this should be constant, no?
+                        is_constant: IsConstant::No,
```

Comment:
Reassignment expressions are always treated as being non-const. But I'm pretty sure we could just take the const-ness of the rhs (?)

Reply:
yeah, but I think const evaluation actually doesn't rely on this after all. cc @vaivaswatha -- this could perhaps be removed entirely.

Reply:
> yeah, but I think const evaluation actually doesn't rely on this after all. cc @vaivaswatha -- this could perhaps be removed entirely.

That's right. Constant evaluation doesn't worry about `is_constant`.

Reply:
And yes, from what I can tell, `is_constant` shouldn't be needed anymore at all.

Reply:
Ok, created #2572 to address this. Therefore we can resolve this conversation as `IsConstant` is dead code.